### PR TITLE
Lor11212/stateless prt wrapper

### PR DESCRIPTION
**Pull request by lor11212**
JIRA: https://zrh-web.esri.com/jira/browse/CE-9592

This PR aims at refactoring the `ModelGenerator` and `RhinoPRTAPI` to make them stateless. This is still a work in progress.

### Implementation

- Removed class variables of `ModelGenerator` and `RhinoPRTAPI` that are used for the generation.
- Removed API functions that were setting rule attribute values to `ModelGenerator`. Instead, attribute values are now encoded into a big map and passed when calling `Generate`.
- Refactored API function `Generate` to take all arguments needed for the generation.
- Refactored API function `Generate` to return meshes, reports and materials in one container object.
- Made the rule attribute extraction stateless by providing a function that extracts rule attributes from the rule.
- Make the default value evaluation stateless as well.

**Remaining work:**
- Add CGA prints and errors to the returned elements of generate.
- Refactor how the flag `doGenerateMaterial` is passed. Add it as an argument of `Generate`.

### CODE REVIEW

**Review by mistafunk — CHANGES_REQUESTED** [review]
Did is a first review pass. I've yet to grok all the details, but already found some smaller issues.

Before I do a second, deeper, pass, I would like to ask you to cleanup the commits a bit, this would help me better understand the changes:
* Please merge "started x" and "finished x" type of commits.
* Avoid committing commented out code
* Avoid putting whitespace cleanups into more important commits.
* See my inline comments.

File: PumaRhino/utils.h [hunk 1]
```diff
@@ -135,6 +140,41 @@ std::basic_string<C>& replace_not_in_range(std::basic_string<C>& str, const std:
 	return str;
 }
 
+/**
+ * Interop helpers
+ */
+
+template<typename T>
+void fillMapBuilder(const std::wstring& key, const T& value, AttributeMapBuilderPtr& aBuilder) {
+	throw std::invalid_argument("Received type is not supported");
```
[the remark this answers is not in the document]
Comment (lor11212) [reply PumaRhino/utils.h]: I tried several things with static_assert but couldn't make it work so far. I probably don't use it correctly. For this example, T should be an int or a double, so something like should work right?
```
template <typename T>
void fillMapBuilder(const std::wstring& /* key */, T /* value */, AttributeMapBuilderPtr& /* aBuilder */) {
	static_assert(std::is_same<T, int>::value || std::is_same<T, double>::value, "Type T should be either int or double");
}
```

File: PumaRhino/PumaGrasshopperAPI.cpp [hunk 2]
```diff
@@ -93,186 +102,351 @@ inline RHINOPRT_API bool AddInitialMesh(ON_SimpleArray<const ON_Mesh*>* pMesh) {
...
-	pcu::appendToRhinoString(*pName, ruleAttr->mFullName);
-	pcu::appendToRhinoString(*pNickname, ruleAttr->mNickname);
-	*type = ruleAttr->mType;
-
-	if (ruleAttr->groups.size() > 0)
-		*pGroup += ON_wString(ruleAttr->groups.front().c_str());
-
-	return true;
-}
+#ifdef DEBUG
```
Comment (mistafunk) [inline PumaRhino/PumaGrasshopperAPI.cpp:179]: Consider introducing a `constexpr bool DBG` and use `if constexpr (DBG)` to query it.
The value of DBG can still be controlled through DEBUG compile definition if needed.

File: PumaRhino/utils.cpp [hunk 3]
```diff
@@ -310,4 +317,164 @@ void PathRemover::operator()(std::filesystem::path const* p) {
...
+
+template <>
+void fillArrayMapBuilder<double>(const std::wstring& key, const std::vector<const wchar_t*>& values,
+                                 AttributeMapBuilderPtr& aBuilder) {
+	double* dArray = new double[values.size()];
+	for (int i = 0; i < values.size(); ++i) {
+		dArray[i] = (double)std::stod(std::wstring(values[i]));
+	}
+	aBuilder->setFloatArray(key.c_str(), dArray, values.size());
+}
```
Comment (mistafunk) [inline PumaRhino/utils.cpp:416]: template functions should be moved to the header file so they are fully visible to the compiler at their instantiation site.

File: PumaRhino/utils.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (mistafunk) [reply PumaRhino/utils.h]: I think you just forgot to allow bool as type as well:

```
static_assert(std::is_same<T, int>::value || std::is_same<T, double>::value || std::is_same<T, bool>::value,
	              "Type T must be one of int, double or bool");
```

**Review by mistafunk — CHANGES_REQUESTED** [review]
Looks good, just a few small issues left.
Quick functional check with the example scene looks good.

**Review by mistafunk — APPROVED** [review]
LGTM, feel free to merge.

**Pull request merged**